frontend/components/PaginatedList/PaginatedList.tsx

```diff
@@ -0,0 +1,279 @@
+    renderItemRow,
+    onToggleItem,
+    onUpdate,
+    isSelected,
+    disabled = false,
+    heading,
+  }: IPaginatedListProps<TItem>,
+  ref: Ref<IPaginatedListHandle<TItem>>
+) {
+  // The # of the page to display.
+  const [currentPage, setCurrentPage] = useState(0);
+  // The set of items fetched via `fetchPage`.
+  const [items, setItems] = useState<TItem[]>([]);
+  // The total # of items fetched via `fetchCount`.
+  const [totalItems, setTotalItems] = useState(0);
+  // The set of items that have been changed in some way.
+  const [dirtyItems, setDirtyItems] = useState<Record<string | number, TItem>>(
+    {}
+  );
+  const [isLoadingPage, setIsLoadingPage] = useState(false);
+  const [isLoadingCount, setIsLoadingCount] = useState(false);
+  const [error, setError] = useState<Error | null>(null);
+  const idKey = _idKey ?? "id";
+  const labelKey = _labelKey ?? "name";
+  const pageSize = _pageSize ?? 20;
+
+  // When the current page # changes, fetch a new page of data.
+  useEffect(() => {
+    let isCancelled = false;
+
+    async function loadPage() {
+      try {
+        setIsLoadingPage(true);
+        setError(null);
+        const result = await fetchPage(currentPage);
+        if (!isCancelled) {
+          setItems(result);
+        }
+      } catch (err) {
+        if (!isCancelled) {
+          setError(err as Error);
+        }
+      } finally {
+        if (!isCancelled) {
+          setIsLoadingPage(false);
+        }
+      }
+    }
+
+    loadPage();
+
+    return () => {
+      isCancelled = true;
+    };
+  }, [currentPage, fetchPage]);
+
+  // Fetch the total # of items.
+  // This will generally only happen once, assuming the parent
+  // uses useCallback() to memoize the fetchCount function.
+  // To retrigger this (for example, after an item is added or removed),
+  // the parent can add dependencies to the useCallback().
+  useEffect(() => {
+    let isCancelled = false;
+
+    async function loadCount() {
+      try {
+        if (!fetchCount) {
+          return;
+        }
+        setIsLoadingCount(true);
+        const result = await fetchCount();
+        if (!isCancelled) {
+          setTotalItems(result);
+        }
+      } catch (err) {
+        if (!isCancelled) {
+          setError(err as Error);
+        }
+      } finally {
+        if (!isCancelled) {
+          setIsLoadingCount(false);
+        }
+      }
+    }
+
+    loadCount();
+
+    return () => {
+      isCancelled = true;
+    };
+  }, [fetchCount]);
+
+  // Whenever the dirty items list changes, notify the parent.
+  useEffect(() => {
+    if (onUpdate) {
+      onUpdate(Object.values(dirtyItems));
+    }
+  }, [onUpdate, dirtyItems]);
+
+  // Create an imperative handle for this component so that parents
+  // can call `ref.current.getDirtyItems()` to get the changed set.
+  useImperativeHandle(ref, () => ({
+    getDirtyItems() {
+      return Object.values(dirtyItems);
+    },
+  }));
+
+  function disableNext() {
```

Comment:
Just heads up our general pattern for functions is 
`const disableNext = () => {` instead of using `function`

Reply:
I don't know if we're moving away from this or not, so let's not change what you got

Reply:
I'll fix this -- `const` is a better pattern as it doesn't get hoisted.  Will fix.